perf: Respect Spark's PARQUET_FILTER_PUSHDOWN_ENABLED config

andygrove opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

N/A

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

My primary motivation was to be able to run benchmarks with the new scans with and without Parquet pushdown enabled, since we know that there are performance issues.

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

Respect Spark's PARQUET_FILTER_PUSHDOWN_ENABLED config.

## How are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

I ran TPC-H @ 100 GB locally with `native_datafusion` scan.

With pushdown enabled: 328 s
With pushdown disabled: 270 s

andygrove commented:
@mbutrovich @parthchandra I am not very familiar with the code that I updated. Could you review?

mbutrovich commented on native/core/src/parquet/parquet_exec.rs:
```diff
@@ -102,10 +104,13 @@ pub(crate) fn init_datasource_exec(
+) -> (TableParquetOptions, SparkParquetOptions) {
     let mut table_parquet_options = TableParquetOptions::new();
+    table_parquet_options.global.pushdown_filters = pushdown_filters;
     // TODO: Maybe these are configs?
```
I think we can remove the TODO, and maybe get rid of the `reorder_filters = true` on the line below. DF still defaults that to false as well so we might not understand the performance implications of it yet. We could add a config for that in a follow up PR and measure the difference.

andygrove replied:
Thanks. I have updated this and also added some notes to the tuning guide.

mbutrovich reviewed (approved):
Other than the one optional suggestion, this LGTM.

mbutrovich commented:
Now that I think about it more: should this be a distinct config from Spark's (i.e., a Comet config like we might do with reorder_filters)? If we fallback to Spark scan are we hurting performance if this is set to false (as suggested by the tuning guide). Maybe we expect if someone is enabling the experimental reader, they should know that their workload is already capable of not falling back to Spark for scan. We can revisit these configs and defaults in the future, I suppose. For now I think this is reasonable for benchmarking the experimental readers.

andygrove commented:
> Now that I think about it more: should this be a distinct config from Spark's

I was also thinking about this. If we add a Comet configuration, we can set the default to disable the filter pushdown for now. I will make this change.

huaxingao commented on spark/src/main/scala/org/apache/comet/parquet/CometParquetFileFormat.scala:
```diff
@@ -96,7 +96,7 @@ class CometParquetFileFormat extends ParquetFileFormat with MetricsSupport with
     val parquetOptions = new ParquetOptions(optionsMap, sqlConf)
     val datetimeRebaseModeInRead = parquetOptions.datetimeRebaseModeInRead
-    val parquetFilterPushDown = sqlConf.parquetFilterPushDown
+    val parquetFilterPushDown = CometConf.COMET_PARQUET_FILTER_PUSHDOWN_ENABLED.get(sqlConf)
```
This shouldn't break iceberg integration.

andygrove replied:
I had to revert this change because it caused failures in the Spark SQL tests

parthchandra commented:
I don't see where `PARQUET_FILTER_PUSHDOWN_ENABLED` is used at all. Wondering how this affects performance. 
Also, row index columns are tied to predicate pushdown. With row indexes, a file reader is allowed to filter row groups based on the predicate but cannot apply the predicate to filter out the values (because the row index generation cannot figure out which row numbers got filtered out).

parthchandra commented:
> I don't see where PARQUET_FILTER_PUSHDOWN_ENABLED is used at all
Nvm. It is used to decide whether to pass in the predicates to the batch reader which uses it to filter out row groups. Note that the behavior of the native code must match this exactly especially for row index generation. So if we want to have a Comet config then it must the same value as the Spark config which makes it redundant.

parthchandra commented on common/src/main/java/org/apache/comet/parquet/Native.java:
```diff
@@ -257,7 +257,8 @@ public static native long initRecordBatchReader(
       String sessionTimezone,
-      int batchSize);
+      int batchSize,
+      boolean pushdownFilters);
```
This is already implied if the filter parameter is null.

andygrove replied:
Thanks. I will update this today.

parthchandra commented on native/core/src/parquet/mod.rs:
```diff
@@ -703,6 +704,7 @@ pub unsafe extern "system" fn Java_org_apache_comet_parquet_Native_initRecordBat
             file_groups,
             None,
             data_filters,
+            pushdown_filters != 0,
```
Same here. `data_filters` is an `Option` and should be `None` if filter pushdown is disabled.

andygrove commented:
I addressed the feedback but I no longer see a performance improvement with `native_datafusion` when disabling `PARQUET_FILTER_PUSHDOWN_ENABLED`, so I have moved this to draft while I figure out why.

andygrove commented:
> I addressed the feedback but I no longer see a performance improvement with `native_datafusion` when disabling `PARQUET_FILTER_PUSHDOWN_ENABLED`, so I have moved this to draft while I figure out why.

I was testing with the wrong Comet JAR file :man_facepalming: 

This is ready for re-review @mbutrovich @parthchandra

andygrove commented on native/core/src/parquet/parquet_exec.rs:
```diff
@@ -61,7 +61,12 @@ pub(crate) fn init_datasource_exec(
     session_timezone: &str,
 ) -> Result<Arc<DataSourceExec>, ExecutionError> {
-    let (table_parquet_options, spark_parquet_options) = get_options(session_timezone);
+    let has_data_filters = match &data_filters {
+        Some(filters) => !filters.is_empty(),
+        None => false,
+    };
```
This may be redundant .. if there are no filters then we probably don't need to change the DataFusion pushdown options

andygrove replied:
I removed this code
